src/source-updater.js

```diff
@@ -24,54 +24,144 @@ const actions = {
   },
   callback: (callback) => (type, sourceUpdater) => {
     callback();
+  },
+  duration: (duration) => (sourceUpdater) => {
+    sourceUpdater.mediaSource.duration = duration;
   }
 };
 
 const updating = (type, sourceUpdater) => {
   const sourceBuffer = sourceUpdater[`${type}Buffer`];
 
-  return (sourceBuffer && sourceBuffer.updating) || sourceUpdater.queue[type].pending;
+  return (sourceBuffer && sourceBuffer.updating) || sourceUpdater.queuePending[type];
 };
 
-const shiftQueue = (type, sourceUpdater) => {
-  const queue = sourceUpdater.queue[type];
+const processMediaSourceQueueEntry = (index, sourceUpdater) => {
+  const queueEntry = sourceUpdater.queue[index];
+  const types = queueEntry.type;
+  const bufferInUse =
+    types.reduce((acc, type) => acc || updating(type, sourceUpdater), false);
 
-  if (updating(type, sourceUpdater) || !queue.actions.length || !sourceUpdater.started_) {
-    return;
+  if (bufferInUse) {
+    // block until all buffers are ready for use
+    return false;
   }
 
-  const action = queue.actions.shift();
+  // remove the entry since it'll be processed
+  sourceUpdater.queue.splice(index, 1);
+  // when the entry needs multiple source buffer types, an individual type doesn't need to
+  // be passed for the action
+  queueEntry.action(sourceUpdater);
+
+  if (queueEntry.doneFn) {
+    queueEntry.doneFn();
+  }
 
-  queue.pending = action[1];
-  action[0](type, sourceUpdater);
+  return true;
 };
 
-const pushQueue = (type, sourceUpdater, action) => {
-  const queue = sourceUpdater.queue[type];
+const nextQueueIndexOfType = (type, queue) => {
+  for (let i = 0; i < queue.length; i++) {
+    const queueEntry = queue[i];
 
-  queue.actions.push(action);
-  shiftQueue(type, sourceUpdater);
+    if (Array.isArray(queueEntry.type)) {
+      // If the next entry that uses this type is a media source entry (uses multiple
+      // source buffers), block processing to allow it to go through first.
+      return null;
+    }
+
+    if (queueEntry.type === type) {
+      return i;
+    }
+  }
+
+  return null;
 };
 
-const onUpdateend = (type, sourceUpdater) => () => {
-  const queue = sourceUpdater.queue[type];
+const shiftQueue = (type, sourceUpdater) => {
+  if (sourceUpdater.queue.length === 0) {
+    return;
+  }
 
-  if (!queue.pending) {
-    shiftQueue(type, sourceUpdater);
+  let queueIndex = 0;
+  let queueEntry = sourceUpdater.queue[queueIndex];
+
+  if (Array.isArray(queueEntry.type)) {
+    // Right now any action which requires multiple source buffer types is considered a
+    // media source action.
+    if (processMediaSourceQueueEntry(queueIndex, sourceUpdater)) {
+      // Only specific source buffer actions must wait for async updateend events. Media
+      // source actions process synchronously. Therefore, we can immediately process the
+      // next items.
+      queueEntry.type.forEach((entryType) => {
+        shiftQueue(entryType, sourceUpdater);
```

Comment:
wouldn't this result in 2 queue actions being process simulataneously?

Reply:
Yup, one for each source buffer type.

Reply:
I would modify the comment to make it clear that we're moving on to the next audio and video type actions. Actually, making it an explicit 2 liner would be good to make things clearer